arlo-client/src/components/Audit/Setup/useAuditSettings.tsx

```diff
@@ -0,0 +1,73 @@
+import { useEffect, useCallback, useState } from 'react'
+import { api, checkAndToast } from '../../utilities'
+import { IAuditSettings, IErrorResponse } from '../../../types'
+
+const defaultValues: IAuditSettings = {
+  state: null,
+  electionName: null,
+  online: null,
+  randomSeed: null,
+  riskLimit: null,
+}
+
+type TNewSettings =
+  | {
+      state: IAuditSettings['state']
+    }
+  | {
+      electionName: IAuditSettings['electionName']
+      online: IAuditSettings['online']
+      randomSeed: IAuditSettings['randomSeed']
+      riskLimit: IAuditSettings['riskLimit']
+    }
+
+const useAuditSettings = (
+  electionId: string
+): [IAuditSettings, (arg0: TNewSettings) => Promise<boolean>] => {
+  const [settings, setSettings] = useState(defaultValues)
+
+  const getSettings = useCallback(async (): Promise<IAuditSettings> => {
+    const settingsOrError: IAuditSettings | IErrorResponse = await api(
+      `/election/${electionId}/settings`
+    )
+    if (checkAndToast(settingsOrError)) {
+      return defaultValues
+    }
+    return settingsOrError
+  }, [electionId])
+
+  const updateSettings = async (
+    newSettings: TNewSettings
+  ): Promise<boolean> => {
+    const oldSettings = await getSettings()
+    const mergedSettings = {
+      ...oldSettings,
+      ...newSettings,
+    }
+    const response: IErrorResponse = await api(
+      `/election/${electionId}/settings`,
+      {
+        method: 'PUT',
+        body: JSON.stringify(mergedSettings),
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      }
+    )
+    if (checkAndToast(response)) {
+      return false
+    }
+    setSettings(mergedSettings)
+    return true
+  }
+
+  useEffect(() => {
+    ;(async () => {
+      const newSettings = await getSettings()
+      setSettings(newSettings)
+    })()
+  }, [getSettings])
+  return [settings, updateSettings]
```

Comment:
if I'm reading this right, I think the settings are fetched once every time a `useAuditSettings` is called, then cached in state until updated. Is that right? If so, I'm good with it, I have a small bit of concern about there being a "hidden" cache here that could bite us later, but I'm not seeing an explicit way in which this would fail, so I can live with it.

Reply:
It looks like, for any given `useAuditSettings` instance, it'll cache the values as long as the election ID does not change. It will only use new values if they are set via `updateSettings` or the whole component that uses `useAuditSettings` is torn down and built again. Values changing on the server would therefore not be reflected here, but maybe that's okay?

Reply:
@benadida @eventualbuddha That's how I've set it up, yes. This is mainly for code reuse, since we're accessing the settings data from two separate forms.

If values change on the server in between the initial loading of the component and submitting values from the form, it will only overwrite the changes in the values represented in the form you're on. That way there are no surprises. Does that make sense?